Main.hs

```
@@ -4,30 +4,117 @@ import Control.Monad (unless)
+  | RootOfNegative Expr
 
 instance Show Error where
-  show = undefined
+  show e = "ERROR! " ++ case e of
+    DivisionByZero expr -> "Cannot divide by zero in \"" ++ show expr ++ "\""
+    RootOfNegative expr -> "Cannot take a root of a negative number in \"" ++ show expr ++ "\""
 
 instance Eq Error where
-  (==) = undefined
+  (==) (DivisionByZero exp1) (DivisionByZero exp2) = exp1 == exp2
+  (==) (RootOfNegative exp1) (RootOfNegative exp2) = exp1 == exp2
+  (==) _ _ = False
+
+evalBinaryOperands :: Expr -> Expr -> Either Error (Double, Double)
+evalBinaryOperands a b =
+  case eval a of
+    Left err -> Left err
+    Right eval_a ->
+      case eval b of
+        Left err -> Left err
+        Right eval_b -> Right (eval_a, eval_b)
 
 eval :: Expr -> Either Error Double
-eval = undefined
+eval (Number x)= Right x
+eval expr@(Sqrt x) =
+  case eval x of
+  Left err -> Left err
+  Right eval_x ->
+    if eval_x >= 0
+      then Right (sqrt eval_x)
+      else Left (RootOfNegative expr)
+eval (Plus a b) =
+  case evalBinaryOperands a b of 
+    Left err -> Left err
+    Right (left, right) -> Right (left + right)
+eval (Minus a b) =
+  case evalBinaryOperands a b of 
+    Left err -> Left err
+    Right (left, right) -> Right (left - right)
+eval (Prod a b) =
+  case evalBinaryOperands a b of 
+    Left err -> Left err
+    Right (left, right) -> Right (left * right)
+eval expr@(Divide a b) =
+  case evalBinaryOperands a b of 
+    Left err -> Left err
+    Right (left, right) -> 
+      if right == 0 
+        then Left (DivisionByZero expr) 
+        else Right (left / right)
+eval expr@(Exp a b) =
+  case evalBinaryOperands a b of 
+    Left err -> Left err
+    Right (left, right) -> 
+      let isWholeNumber x = x == fromIntegral (round x) in
+      if left == 0 && right < 0 
+        then Left (DivisionByZero expr) 
+      else if left < 0 && not (isWholeNumber right)
+        then Left (RootOfNegative expr)
+        else Right (left ** right)
```

This is a mess. Throwing a division by zero error from the exponentiation function is weird. If you want to report an error in this case, add another constructor to the `Error` data type.